nurse_scheduling.py

```diff
@@ -0,0 +1,162 @@
+from collections import defaultdict
+from copy import deepcopy
+
+# count nurses n = 1 ... n_nurses
+# count scheduling days as d = 1 ... n_days
+# binary variable q_nd is the assignment of nurse n to day d
+# a is a positive correlation coefficient for implementing the hard nurse
+# constraint - value provided by Ikeda, Nakamura, Humble
+a = 3.5
+n_nurses = 3
+n_days = 11
+size = n_days * n_nurses
+
+# Hard shift constraint: at least one nurse working every day
+# Lagrange parameter, for hard nurse constraint, on workforce and effort
+# Workforce function W(d) - set to a constant 'workforce' for now
+# Effort function E(n) - set to a constant 'effort' for now
+lagrange_parameter = 1.3
+workforce = 1
+effort = 1
+
+# Soft nurse constraint: all nurses should have approximately even work
+#                        schedules
+# Lagrange parameter, for shift constraints, on work days
+# preference function 'preference' - set to a constant for now
+# Minimum duty days 'min_duty_days' - the number of work days that each
+# nurse wants
+# to be scheduled. At present, each will do the minimum on average.
+# The parameter gamma's value suggested by Ikeda, Nakamura, Humble
+gamma = 0.3
+preference = 1
+min_duty_days = int(n_days/n_nurses)
+
+
+# Find index into 1D list for (nurse_index, day_index)
+def get_index(nurse_index, day_index):
+    return nurse_index * n_days + day_index
+
+
+# Inverse of get_index - given an index in a 1D list, return the nurse_index
+# and day_index
+def get_nurse_and_day(index):
+    nurse_index, day_index = divmod(index, n_days)
+    return nurse_index, day_index
+
+
+# Hard nurse constraint: no nurse works two consecutive days
+# It does not have Lagrange parameter - instead, J matrix
+# symmetric, real-valued interaction matrix J, whereas all terms are
+# a or zero.
+# composite indices i(n, d) and j(n, d) as functions of n and d
+# In other words, the indices are not n and d. They are kind of
+# (n+d)**2, and we need to take that into account. This seems to be
+# a clean way to do it.
+# J_i(n,d)j(n,d+1) = a and 0 otherwise.
+J = defaultdict(int)
+for nurse in range(n_nurses):
+    for day in range(n_days - 1):
+        nurse_day_1 = get_index(nurse, day)
+        nurse_day_2 = get_index(nurse, day+1)
+        J[nurse_day_1, nurse_day_2] = a
+
+# Q matrix assign the cost term, the J matrix
+Q = deepcopy(J)
+
+# Hard shift constraint. The sum is over each day.
+# lagrange_parameter * ((sum(effort * q) - workforce) ** 2)
+for nurse_day_1 in range(size):
+    _, date_index = get_nurse_and_day(nurse_day_1)
+    # Diagonal term, without the workforce * workforce
+    Q[nurse_day_1, nurse_day_1] += lagrange_parameter * (1 - (2 * workforce))
+    for nurse_day_2 in range(size):
+        _, day_index_2 = get_nurse_and_day(nurse_day_2)
+        # Include only the same day, across nurses
+        if (date_index == day_index_2 and nurse_day_2 != nurse_day_1):
+            Q[nurse_day_1, nurse_day_2] += lagrange_parameter * 2
```

Comment:
@melody, I'm nervous about this one, because I haven't convinced myself that it is mathematically the same. I used the list approach because I wanted to simplify the loops. If it's OK with you, I'd like to leave it this way. If you're feeling strongly, I'll work through it carefully, but in the interests of time, I'd prefer not to.